src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1472,53 +1452,76 @@ public IHtmlContent GenerateGroupsAndOptions(string optionLabel, IEnumerable<Sel
+            var optionGenerated = new bool[itemsList.Count];
+            for (var i = 0; i < itemsList.Count; i++)
+            {
+                if (!optionGenerated[i])
                 {
-                    var groupBuilder = new TagBuilder("optgroup");
-                    if (optGroup.Name != null)
+                    var item = itemsList[i];
+                    var optGroup = item.Group;
+                    if (optGroup != null)
                     {
-                        groupBuilder.MergeAttribute("label", optGroup.Name);
-                    }
+                        var groupBuilder = new TagBuilder("optgroup");
+                        if (optGroup.Name != null)
+                        {
+                            groupBuilder.MergeAttribute("label", optGroup.Name);
+                        }
 
-                    if (optGroup.Disabled)
-                    {
-                        groupBuilder.MergeAttribute("disabled", "disabled");
-                    }
+                        if (optGroup.Disabled)
+                        {
+                            groupBuilder.MergeAttribute("disabled", "disabled");
+                        }
 
-                    groupBuilder.InnerHtml.AppendLine();
-                    foreach (var item in group)
-                    {
-                        groupBuilder.InnerHtml.AppendLine(GenerateOption(item));
-                    }
+                        groupBuilder.InnerHtml.AppendLine();
 
-                    listItemBuilder.AppendLine(groupBuilder);
-                }
-                else
-                {
-                    foreach (var item in group)
+                        for (var j = i; j < itemsList.Count; j++)
+                        {
+                            var groupItem = itemsList[j];
+
+                            if (!optionGenerated[j] &&
+                                groupItem.Group != null &&
+                                groupItem.Group.GetHashCode() == item.Group.GetHashCode())
+                            {
+                                groupBuilder.InnerHtml.AppendLine(GenerateOption(groupItem, currentValues));
+                                optionGenerated[j] = true;
+                            }
+                        }
+
+                        listItemBuilder.AppendLine(groupBuilder);
+                    }
+                    else
                     {
-                        listItemBuilder.AppendLine(GenerateOption(item));
+                        listItemBuilder.AppendLine(GenerateOption(item, currentValues));
+                        optionGenerated[i] = true;
                     }
                 }
             }
 
             return listItemBuilder;
         }
 
-        private IHtmlContent GenerateOption(SelectListItem item)
+        private IHtmlContent GenerateOption(SelectListItem item, ICollection<string> currentValues)
```

Comment:
`currentValues` should be either `HashSet<string>` or `ISet<string>` to avoid Linq in this method.